### DISCUSSION TITLE
Recover the old voice state in VoiceStateUpdateEvent

### DISCUSSION THREAD

**Discussion by EnoPM** [diamondburned/arikawa#352] [Help] [answered]
Hello,
I must have missed something but I couldn't find a way to retrieve the old voice state in a VoiceStateUpdate handler. The event returns the new state but the cache is already updated when it is called
Thank you in advance for your help.

**Comment by samhza** [comment, marked as answer]
I believe `PreHandler` is what you're looking for. It's executed before the cache is updated, so you'll be able to get the old voice state. The [undeleter](https://github.com/diamondburned/arikawa/blob/v3/0-examples/undeleter/main.go) example shows how it can be used.

**Reply by EnoPM** [reply]
Thanks that sounds like what I need. However *State.PreHandler is nil by default. I have to change something in the configuration?

**Reply by samhza** [reply]
Use [`handler.New()`](https://pkg.go.dev/github.com/diamondburned/arikawa/v3/utils/handler#New) to create a new handler.

**Comment by EnoPM** [comment]
If someone has the same question, you should know that to be sure that the cache contains the old state, the handler must be synchronous.

**Comment by diamondburned** [comment]
There's a good example of this in [undeleter](https://github.com/diamondburned/arikawa/blob/v3/0-examples/undeleter/main.go#L24).
